### ISSUE TITLE
[Bounty $1000] moreh_adam / moreh_adamw silently run different hyperparameters than requested: bf16 truncation turns beta2=0.999 into 0.99609 (V-horizon 1000 -> 256 steps), and 0.9999 is indistinguishable from 0.999

### ISSUE THREAD

**Issue by natesmind** [tenstorrent/tt-metal#55935]
## Summary

With the default compute config, `moreh_adam` and `moreh_adamw` truncate all
five optimizer hyperparameters (`lr`, `beta1`, `beta2`, `eps`, `weight_decay`)
to bfloat16 before any math happens. Three consequences:

1. **The optimizer that runs is not the optimizer that was requested.**
   `beta2 = 0.999` executes as `0.99609375`: the second-moment averaging
   horizon `1/(1-beta2)` collapses from 1,000 steps to 256. `beta2 = 0.9999`
   truncates to the *same* value — a source-level replay shows the two settings
   produce bit-identical trajectories (torch's differ from step 1).
2. **`moreh_adamw` is internally inconsistent**: its EMA coefficients
   `(1 - beta)` are computed on device from the *truncated* betas, while its
   bias corrections come from host `std::pow(beta_exact, step)`. The moment
   accumulators and their corrections assume different betas; in a source-level
   replay the parameter trajectory deviates from `torch.optim.AdamW` by
   **34-44%** through steps 10-100 (exact-arithmetic replay; only the scalars
   are truncated exactly as the reader kernel does).
3. **PCC-based tests are structurally blind to this**: a wrong beta mostly
   rescales trajectories, and PCC is scale-invariant. (Separately, the nightly
   test's param assert is overwritten before it fires, as already documented in
   #46776.)

This is disjoint from #46776 (param_out write-back garbage — a kernel bug that
reproduces with correct scalars), from #54709 (fill-loop *performance*), and
from #51278 (CB allocations).

## Mechanism

The reader delivers scalars by bit-truncation, not conversion
([`reader_moreh_adam.cpp#L13-L30`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/reader_moreh_adam.cpp#L13-L30)):

```cpp
#else
    CoreLocalMem<uint16_t> ptr(dfb.get_write_ptr());
    for (int j = 0; j < 1024; j++) {
        ptr[j] = uint16_t(value >> 16);   // fp32 -> bf16 by truncation
    }
#endif
```

The scalar CB is bf16 whenever fp32 dest-acc is off
([`moreh_adam_program_factory.cpp#L75`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_program_factory.cpp#L75)):

```cpp
auto intermed_cb_format = fp32_dest_acc_en ? tt::DataFormat::Float32 : data_format;
```

and fp32 dest-acc is off by default
([`compute_kernel_config.hpp#L41-L48`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.hpp#L41-L48):
`default_fp32_acc = false`; `moreh_adam` passes only a fidelity override at
[`moreh_adam_device_operation.cpp#L145`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp#L145)).
Params are restricted to BFLOAT16 by validation
([`moreh_helper_functions.hpp#L153`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/moreh/moreh_helper_functions.hpp#L153)),
so every legal call without an explicit fp32 config takes this path.

The kernel then computes `(1 - beta)` and (in `moreh_adam`) `beta^step` from the
truncated tiles
([`moreh_adam.cpp#L104-L174`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/kernels/moreh_adam.cpp#L104-L174)),
recomputing `beta1^step` and `beta2^step` with `power_tile` **inside the
per-tile loop** (L146-L157 and L254-L265) even though both are constants for
the whole op.

## Quantified impact

Truncation of the torch-default hyperparameters:

| scalar | requested | delivered | rel. err | EMA horizon |
|--------|-----------|-----------|----------|-------------|
| lr | 1e-3 | 9.9945e-4 | 0.055% | — |
| beta1 | 0.9 | 0.8984375 | 0.17% | 10 -> 9.8 |
| **beta2** | **0.999** | **0.99609375** | 0.29% | **1,000 -> 256** |
| **beta2** | **0.9999** | **0.99609375** | 0.38% | **10,000 -> 256** |
| eps | 1e-8 | 9.954e-9 | 0.47% | — |
| weight_decay | 1e-2 | 9.949e-3 | 0.51% | — |

- `beta2 = 0.999` and `beta2 = 0.9999` deliver **bit-identical** trajectories
  (max |difference| = 0 over 100 replayed steps; torch's trajectories differ by
  2.5e-5 at step 100 on the same problem). Every beta2 in [0.99609375, 1) is
  the same optimizer.
- `moreh_adamw` replay vs `torch.optim.AdamW` (quadratic objective, wd = 0.01,
  scalars truncated exactly as the reader does, all tensor math exact fp32 to
  isolate the effect): parameter deviation 34% at step 10, 44% at step 100.
- `moreh_adam` (self-consistent truncation) behaves as a *correct* Adam whose
  beta2 is 0.99609 — earlier-step deviations are small on smooth objectives but
  the second-moment estimator reacts ~4x faster than requested, which is
  precisely what beta2 exists to control on noisy gradients.

## A trap for the fix

Round-to-nearest would be worse, not better: bf16 RTN of `0.999` is exactly
`1.0`, making `1/(1 - beta2^step)` infinite. The scalars must reach the kernel
in fp32 — rounding mode is not the issue.

## Proposed fix

`moreh_adamw` already computes `std::pow(beta, step)` on host
([`multi_core_program_factory.cpp#L282-L283`](https://github.com/tenstorrent/tt-metal/blob/1867f508f94ca072ccde1a9afbde09a5ec57e1f2/ttnn/cpp/ttnn/operations/moreh/moreh_adamw/device/multi_core_program_factory.cpp#L282-L283)) —
extend that design:

1. Host computes, in fp64/fp32: `one_minus_beta1`, `one_minus_beta2`,
   `step_size = lr / (1 - beta1^step)`, `inv_bias_correction2
   = 1 / (1 - beta2^step)` — from the exact scalars.
2. Deliver scalars in a CB that is **always Float32** (they are op constants;
   their format need not follow the param dtype), with the reader writing
   uint32 unconditionally. Dest-register precision for tile math is unchanged;
   what matters is that coefficient *values* are right.
3. Delete the in-kernel `power_tile` bias-correction blocks in `moreh_adam` —
   also removing a per-tile recomputation of two step-dependent powers, a
   side-benefit adjacent to (but distinct from) the fill-cost issue in #54709.
4. Tests: assert `beta2 = 0.999` vs `0.9999` produce *different* `param_out`
   (bit-identity today), and compare `exp_avg_sq` against torch by value
   (rtol), not PCC, at steps {1, 10, 100} — PCC cannot see coefficient errors.

## Reproduction (host-only, no card required)

```python
import struct, torch

def bf16_trunc(x):  # fill_cb_with_value semantics: fp32 bits >> 16
    b, = struct.unpack("<I", struct.pack("<f", float(x)))
    return struct.unpack("<f", struct.pack("<I", b & 0xFFFF0000))[0]

print(bf16_trunc(0.999), bf16_trunc(0.9999))   # 0.99609375 0.99609375
print(bf16_trunc(0.999) == bf16_trunc(0.9999)) # True

def tt_adam(steps, b2):
    lr, b1, eps = bf16_trunc(1e-3), bf16_trunc(0.9), bf16_trunc(1e-8)
    b2 = bf16_trunc(b2); x, m, v = 1.0, 0.0, 0.0
    for t in range(1, steps + 1):
        g = x
        m = b1 * m + (1 - b1) * g
        v = b2 * v + (1 - b2) * g * g
        x -= (lr / (1 - b1 ** t)) * m / ((v / (1 - b2 ** t)) ** 0.5 + eps)
    return x

print(tt_adam(100, 0.999) == tt_adam(100, 0.9999))  # True: same optimizer

def torch_adam(steps, b2):
    x = torch.tensor([1.0], requires_grad=True)
    opt = torch.optim.Adam([x], lr=1e-3, betas=(0.9, b2))
    for _ in range(steps):
        opt.zero_grad(); (0.5 * x * x).sum().backward(); opt.step()
    return x.item()

print(abs(torch_adam(100, 0.999) - torch_adam(100, 0.9999)))  # ~2.5e-05: different
```

(A fuller replay including the adamw EMA/bias-correction inconsistency and the
34-44% divergence table is a ~60-line extension of the same script.)

Caveat: I do not have a card; this is a source-level replay of the scalar path
(the quoted reader/factory/kernel code), which is where the defect lives. The
tensor-side arithmetic is modeled as exact to isolate the scalar effect; on
device, bf16 tile math adds further error on top.

Note: suggested under the community-suggested-bounty process; proposed tier:
medium. I'd like to take the implementation if approved — happy to be assigned.

**Comment by mateusznowakTT** [comment]
@natesmind : For bounty proposals please first submit it to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com) for review.
